float32 angle
bool defined